[inductor] Use shape env bounds in inductor bounds.py (#106175)

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106568

Summary: If constrained ranges are available, use them in bounds.py before value range analysis (to enable Int64 -> Int32 optimization).

Test Plan: New unit test in test_torchinductor.py to mark a tensor as dynamic, then constrain with constrain_as_size (as outlined in https://github.com/pytorch/pytorch/issues/106175)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

masnesral committed:
[inductor] Use shape env bounds in inductor bounds.py (#106175)

Summary: If constrained ranges are available, use them in bounds.py before value range analysis (to enable Int64 -> Int32 optimization).

Test Plan: New unit test in test_torchinductor.py to mark a tensor as dynamic, then constrain with constrain_as_size (as outlined in https://github.com/pytorch/pytorch/issues/106175)

[ghstack-poisoned]

eellison commented on torch/_inductor/bounds.py:
```diff
@@ -22,12 +22,22 @@ class BoundVars:
+        tracing_context = torch._guards.TracingContext.get()
+        if tracing_context:
+            subs = [
+                (k, v.upper)
```
we should have separate substitutions for upper and lower bound

masnesral replied:
@eellison - thanks; meant to ask you about the lower bound handling. I might not understand exactly what we're doing here. If we have a range that's a symbolic expression of `4 * s0`, for example, but we also have a constrained `var_range` available like `s0: ValueRanges(lower=2, upper=100)` ... I guess I assumed that only the upper bound was relevant for the int64 -> int32 optimization and we could use `4 * 100 = 400` for the upper bound. How is the lower bound relevant exactly? Do you just mean that we can also further constrain the lower bound on this line (and not assume 0)?

`   ValueRanges(0, s - 1)`

lezcano replied:
If `x` has bounds `[a, b]`, then `-x` has bounds `[-b, -a]`, so you need to track both.

eellison commented on torch/_inductor/lowering.py:
```diff
@@ -4520,6 +4520,11 @@ def fn(*args, **kwargs):
 
 
+@register_lowering(aten.sym_constrain_range)
+def sym_constrain_range(a, min, max):
```
should we assert that `a` is either an int or if its a symint that it's found in the tracing context ?

masnesral replied:
Maybe I misunderstand, but at least in my testing `a` is neither an int nor a symint. It's a sympy Symbol. But I can check that it's in the tracing context.

eellison commented on test/inductor/test_torchinductor.py:
```diff
@@ -6979,6 +6979,38 @@ def fn(x: torch.Tensor) -> torch.Tensor:
+                return a[y.to(torch.int64)]
+
+            def fn2(a: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
+                from torch._export.constraints import constrain_as_size
```
nit, this can be a top level import

eellison reviewed (commented):
awesome 🚀 I have a couple comments, but looks good!

lezcano commented on torch/_inductor/bounds.py:
```diff
@@ -22,12 +22,22 @@ class BoundVars:
+            ]
+        self.replacement_vals = {}
+        for k, v in loop_body.var_ranges.items():
+            s = v.subs(subs)
```
Is `var_ranges` just a symint, or can it be a expression? If it can be an expression, you should use `bound_sympy` to compute the actual bounds given the `subs`. If it's not, we can simply look up `v` in `subs` and get its range if it has `free_symbols`.

masnesral replied:
@lezcano (Pardon my being a noob here). `loop_body.var_ranges` can contain expressions, e.g., I see `4 * s0`. Given that, are you simply suggesting to use the `bound_sympy` helper to evaluate the bounds instead of the `subs` method I've used here? Playing around with that helper function, it seems like it indeed probably does what I want. But it sounds like you anticipate some modifications are needed in that helper first? If so, I don't understand that part.

lezcano replied:
Have a look at how `bound_sympy` is implemented. Before calling into the bulk of the logic, it tries to do something very similar to what you're doing here. Just move the logic you implemented here to `bound_sympy` and you should be able to avoid repeating it :)

masnesral replied:
Thanks. I might be naive, but looks like `bound_sympy` might "just work". I'll put up a new rev to show you what I mean.

lezcano reviewed (commented):
The same should be done in `bound_sympy`, as that will also affect the bounds we use to elide guards.

Even more, if `var_ranges` can be an expression, then we can probably just implement this in `bound_sympy` and delegate the calculation of the bounds in `replacement_vars` to `bound_sympy`

masnesral committed:
@lezcano's comments on "[inductor] Use shape env bounds in inductor bounds.py (#106175)"


Summary: If constrained ranges are available, use them in bounds.py before value range analysis (to enable Int64 -> Int32 optimization).

Test Plan: New unit test in test_torchinductor.py to mark a tensor as dynamic, then constrain with constrain_as_size (as outlined in https://github.com/pytorch/pytorch/issues/106175)

[ghstack-poisoned]

lezcano commented on torch/_inductor/bounds.py:
```diff
@@ -22,12 +22,19 @@ class BoundVars:
+
+        # For the replacement vals, include constrained ranges from the tracing context
+        self.replacement_vals = {}
+        context = torch._guards.TracingContext.get()
+        ranges = context.fake_mode.shape_env.var_to_range if context else {}
```
We want to move this to `bound_sympy` so that other users of `bound_sympy` also get these better bounds for free.

masnesral replied:
@lezcano I don't understand this code base (obviously), so I just want to extra clarify: Are you sure that's always appropriate? Purely from an API perspective, it seems a little strange to silently add new ValueRanges beyond what were explicitly provided by the caller. Do all usages of `bound_sympy` want these ranges from the tracing context included?

eellison replied:
The tracing context should only be set if its valid to use. If it is set, we can use it.

masnesral committed:
Update on "[inductor] Use shape env bounds in inductor bounds.py (#106175)"


Summary: If constrained ranges are available, use them in bounds.py before value range analysis (to enable Int64 -> Int32 optimization).

Test Plan: New unit test in test_torchinductor.py to mark a tensor as dynamic, then constrain with constrain_as_size (as outlined in https://github.com/pytorch/pytorch/issues/106175)

[ghstack-poisoned]

eellison reviewed (commented):
has some compilation time issues but I think those may already exist, so maybe it's worth addressing those in a follow up. will defer to @lezcano

lezcano commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -545,7 +545,15 @@ def __getattr__(self, name):
+
+    # If there's a tracing context, augment available constrained ranges.
+    context = torch._guards.TracingContext.get()
+    if context:
```
I don't know what you mean by "calculating the bounds happens recursively". As far as I can tell, I don't think it does.

Something that can be done better is avoiding to do the `deepcopy`, and simply merging the dictionaries as `{**x, **y}`. This would avoid a few unnecessary copies. Same same for the `unbounded_vars` bit, where we can avoid creating a new dictionary if we created it here.

eellison replied:
>  Recursive case
args = [sympy_interp(analysis, env, arg) for arg in expr.args]  # type: ignore[arg-type]
handler_name = handlers()[expr.func]
handler = getattr(analysis, handler_name)

We're recursively computing the bounds of symbols without storing them in the environment.

lezcano replied:
But that doesn't call `bound_sympy`

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -4520,6 +4520,14 @@ def fn(*args, **kwargs):
 
 
+@register_lowering(aten.sym_constrain_range)
+def sym_constrain_range(a, min, max):
```
This is not used?

masnesral replied:
@lezcano , I believe this part is needed. Otherwise we get:

```
...
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
MissingOperatorWithoutDecomp: missing lowering
  target: aten.sym_constrain_range.default
```

lezcano reviewed (commented):
Just two small comments, otherwise it LGTM

masnesral committed:
remove deepcopies on "[inductor] Use shape env bounds in inductor bounds.py (#106175)"


Summary: If constrained ranges are available, use them in bounds.py before value range analysis (to enable Int64 -> Int32 optimization).

Test Plan: New unit test in test_torchinductor.py to mark a tensor as dynamic, then constrain with constrain_as_size (as outlined in https://github.com/pytorch/pytorch/issues/106175)

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Use shape env bounds in inductor bounds.py (#106175)"


Summary: If constrained ranges are available, use them in bounds.py before value range analysis (to enable Int64 -> Int32 optimization).

Test Plan: New unit test in test_torchinductor.py to mark a tensor as dynamic, then constrain with constrain_as_size (as outlined in https://github.com/pytorch/pytorch/issues/106175)

[ghstack-poisoned]

lezcano reviewed (approved):
Nice addition! Thank you!

masnesral committed:
fix shape_env is None on "[inductor] Use shape env bounds in inductor bounds.py (#106175)"


Summary: If constrained ranges are available, use them in bounds.py before value range analysis (to enable Int64 -> Int32 optimization).

Test Plan: New unit test in test_torchinductor.py to mark a tensor as dynamic, then constrain with constrain_as_size (as outlined in https://github.com/pytorch/pytorch/issues/106175)

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Use shape env bounds in inductor bounds.py (#106175)"


Summary: If constrained ranges are available, use them in bounds.py before value range analysis (to enable Int64 -> Int32 optimization).

Test Plan: New unit test in test_torchinductor.py to mark a tensor as dynamic, then constrain with constrain_as_size (as outlined in https://github.com/pytorch/pytorch/issues/106175)

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
